packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart

```diff
@@ -0,0 +1,111 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:file/file.dart';
+import 'package:file/memory.dart';
+import 'package:flutter_tools/src/android/migrations/top_level_gradle_build_file_migration.dart';
+import 'package:flutter_tools/src/base/logger.dart';
+import 'package:flutter_tools/src/base/project_migrator.dart';
+import 'package:flutter_tools/src/project.dart';
+import 'package:test/fake.dart';
+
+import '../../src/common.dart';
+import '../../src/context.dart';
+
+void main() {
+  group('Android migration', () {
+    
+    testWithoutContext('migrators succeed', () {
+      final FakeAndroidMigrator fakeAndroidMigrator = FakeAndroidMigrator();
+      final ProjectMigration migration = ProjectMigration(<ProjectMigrator>[fakeAndroidMigrator]);
+      migration.run();
+    });
+
+    group('migrate the Gradle "clean" task to lazy declaration', () {
+      late MemoryFileSystem memoryFileSystem;
+      late BufferLogger bufferLogger;
+      late FakeAndroidProject project;
+      late File topLevelGradleBuildFile;
+
+      setUp(() {
+        memoryFileSystem = MemoryFileSystem();
+        bufferLogger = BufferLogger.test();
+        project = FakeAndroidProject();
+        project.hostAppGradleRoot = memoryFileSystem.currentDirectory;
+        topLevelGradleBuildFile = project.hostAppGradleRoot.childFile('build.gradle');
+      });
+
+      testUsingContext('skipped if files are missing', () {
+        final TopLevelGradleBuildFileMigration androidProjectMigration = TopLevelGradleBuildFileMigration(
+          project,
+          bufferLogger,
+        );
+        androidProjectMigration.migrate();
+        expect(topLevelGradleBuildFile.existsSync(), isFalse);
+
+        expect(bufferLogger.traceText, contains('Top-level Gradle build file not found, skipping "clean" task migration.'));
+        expect(testLogger.statusText, isEmpty);
```

Comment:
Huh, there are two different loggers involved here. That feels fragile — like it'd be easy for a test to wrongly pass because, for example, an `isEmpty` check was made on one logger, but the output it was meant to detect actually appeared on the other logger.

It looks like in `ios_project_migration_test.dart` there's just one logger, and this global `testLogger` isn't used. Having just one logger sounds like a good idea to me.

Reply:
Yup, good catch as well. I removed all these unnecessary excepts.

Reply:
Hmm, `ios_project_migration_test.dart` does have expectations like these on `statusText`, though, so that makes me think they're likely useful. (The difference is they're using the same logger that was passed to the migrator.)

Looking into those, in addition to checking it's empty in some tests, in other tests it has checks like this:
```dart
        expect(testLogger.statusText, contains('Upgrading project.pbxproj'));
```

That in turn looks like a reference to this in the [ProjectMigrator] logic:
```dart
    if (migrationRequired) {
      logger.printStatus('Upgrading $basename');
```

So effectively it's a way of checking that the file was, or wasn't, modified.

In these tests, there's the `lastModifiedSync` checks for not being modified, and when it is modified the change in contents is clear enough. So I'm happy with leaving them out.